pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,179 @@
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &userAccounts); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, u := range userAccounts {
+			u.Type = UserType
+			accounts = append(accounts, u)
+		}
+	}
+	if principalType == "" || principalType == GroupType {
+		var groups []Group
+		searchURL := fmt.Sprintf("%s/%ss?search=%s", sURL, GroupType, searchTerm)
+		search := URLEncoded(searchURL)
+
+		b, err := k.getFromKeyCloak(accessToken, search, config.Certificate, config.PrivateKey)
+		if err != nil {
+			// no match on search returns an error. do not log
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &groups); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, g := range groups {
+			accounts = append(accounts, Account{ID: g.ID, Name: g.Name, Type: GroupType})
+			subGroups := getSubGroups(g)
+			for _, sg := range subGroups {
+				accounts = append(accounts, Account{ID: sg.ID, Name: sg.Name, Type: GroupType})
+			}
+		}
+	}
+	return accounts, nil
+}
+
+func getSubGroups(group Group) []Group {
+	var groups []Group
+	if len(group.Subgroups) > 0 {
+		for i, sub := range group.Subgroups {
+			// setting an upper limit for how many subgroups we will loop through
+			// this value was chosen at random so can be changed if needed
+			if i < 100 {
+				groups = append(groups, sub)
+				groups = append(groups, getSubGroups(sub)...)
+			}
+		}
+	}
+	return groups
+}
+
+func (k *KClient) getFromKeyCloakByID(principalID, accessToken, searchType string, config *v32.OIDCConfig) (Account, error) {
+	sURL, err := k.getSearchURL(config.Issuer)
+	if err != nil {
+		return Account{}, nil
+	}
+	searchURL := fmt.Sprintf("%s/%s/%s", sURL, searchType, principalID)
+	search := URLEncoded(searchURL)
+	b, err := k.getFromKeyCloak(accessToken, search, config.Certificate, config.PrivateKey)
+	if err != nil {
+		return Account{}, nil
+	}
+	var searchResult Account
+	if err := json.Unmarshal(b, &searchResult); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResult, err
+	}
+	return searchResult, nil
+
+}
+func (k *KClient) getSearchURL(issuer string) (string, error) {
+	splitIssuer := strings.SplitAfter(issuer, "/auth/")
+	return fmt.Sprintf(
+		"%sadmin/%s",
+		splitIssuer[0],
+		splitIssuer[1],
+	), nil
+}
+
+//URLEncoded encodes the string
+func URLEncoded(str string) string {
+	u, err := url.Parse(str)
+	if err != nil {
+		logrus.Errorf("Error encoding the url: %s, error: %v", str, err)
+		return str
+	}
+	return u.String()
+}
+
+func (k *KClient) getFromKeyCloak(accessToken, url, certificate, key string) ([]byte, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	err = common.GetClientWithCertKey(k.httpClient, certificate, key)
```

Comment:
Can we do this one time when we initialize `KClient` with the assumption that login would still need to do this

Reply:
because we need the configuration information, we are not able to do it as part of the init, but I did move this code so that it is only happening in twice in the keycloak_provider code - once for search principals and once for get principals.